packages/core/src/components/Form.tsx

```diff
@@ -508,49 +520,70 @@ export default class Form<
 
     const mustValidate = !noValidate && liveValidate;
     let state: Partial<FormState<T, S, F>> = { formData, schema };
-    let newFormData = formData;
 
     if (omitExtraData === true && liveOmit === true) {
       const retrievedSchema = schemaUtils.retrieveSchema(schema, formData);
       const pathSchema = schemaUtils.toPathSchema(retrievedSchema, '', formData);
 
       const fieldNames = this.getFieldNames(pathSchema, formData);
 
-      newFormData = this.getUsedFormData(formData, fieldNames);
       state = {
-        formData: newFormData,
+        formData: this.getUsedFormData(formData, fieldNames),
       };
     }
 
-    if (mustValidate) {
-      const schemaValidation = this.validate(newFormData);
-      let errors = schemaValidation.errors;
-      let errorSchema = schemaValidation.errorSchema;
-      const schemaValidationErrors = errors;
-      const schemaValidationErrorSchema = errorSchema;
-      if (extraErrors) {
-        const merged = validationDataMerge(schemaValidation, extraErrors);
-        errorSchema = merged.errorSchema;
-        errors = merged.errors;
-      }
-      state = {
-        formData: newFormData,
-        errors,
-        errorSchema,
-        schemaValidationErrors,
-        schemaValidationErrorSchema,
-      };
-    } else if (!noValidate && newErrorSchema) {
+    if (!noValidate && newErrorSchema) {
       const errorSchema = extraErrors
         ? (mergeObjects(newErrorSchema, extraErrors, 'preventDuplicates') as ErrorSchema<T>)
         : newErrorSchema;
+
       state = {
-        formData: newFormData,
+        formData,
         errorSchema: errorSchema,
         errors: toErrorList(errorSchema),
       };
     }
-    this.setState(state as FormState<T, S, F>, () => onChange && onChange({ ...this.state, ...state }, id));
+
+    this.setState(state as FormState<T, S, F>, () => {
+      onChange && onChange({ ...this.state, ...state }, id);
+
+      if (mustValidate) {
+        // @ts-ignore
+        const debounceThreshold = liveValidate?.debounceThreshold;
+        const callbackAfterValidation = () => this.props.onChange?.(this.state, id);
+        this.validateAndUpdateStateDebounced(debounceThreshold)(state.formData, callbackAfterValidation);
+      }
+    });
+  };
+
+  validateAndUpdateStateDebounced = _memoize((debounceThreshold: number | null | undefined) => {
```

Comment:
Does it make more sense to use the react version of memoize? or possibly create an instance variable for the debounce function and only change it when the debounceThreshold changes/is removed? Then you create the function in the constructor and in the componentDidUpdate

Reply:
I'm not sure what you mean by "he react version of memoize", but using a single-line cache memoize implementation (something like [memoize-one](https://github.com/alexreardon/memoize-one)) would probably lead to a easier-to-read code :)

Reply:
I mean `React.useMemo()`